src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,281 @@
+                                QueueWaitCompletion(registeredHandle, false);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        for (int i = 0; i < _numUserWaits; i++)
+                        {
+                            RegisteredWaitHandle registeredHandle = _registeredWaitHandles[i];
+                            if (registeredHandle.Timeout == _currentTimeout)
+                            {
+                                QueueWaitCompletion(registeredHandle, true);
+                            }
+                        }
+                    }
+                }
+            }
+
+            private void ProcessRemovals()
+            {
+                _removesLock.Acquire();
+                _registeredHandlesLock.Acquire();
+                if(_numPendingRemoves == 0)
+                {
+                    _registeredHandlesLock.Release();
+                    _removesLock.Release();
+                    return;
+                }
+
+                // This is O(N^2), but max(N) = 63 and N will usually be very low
+                for (int i = 0; i < _numPendingRemoves; i++)
+                {
+                    for (int j = 0; j < _numUserWaits; j++)
+                    {
+                        if (_pendingRemoves[i] == _registeredWaitHandles[j])
+                        {
+                            _registeredWaitHandles[j] = null;
+                            _waitHandles[j + 1] = null;
+                            break;
+                        }
+                    }
+                    _pendingRemoves[i] = null;
+                }
+                _numPendingRemoves = 0;
+
+                // Fill in nulls
+                // This is O(1), Goes through each of the 63 possible handles once.
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (_registeredWaitHandles[i] == null)
+                    {
+                        for (int j = _numUserWaits - 1; j > i; j--)
+                        {
+                            if(_registeredWaitHandles[j] != null)
+                            {
+                                _registeredWaitHandles[i] = _registeredWaitHandles[j];
+                                _registeredWaitHandles[j] = null;
+                                _numUserWaits = j;
+                                break;
+                            }
+                        }
+                        if (_registeredWaitHandles[i] == null)
+                        {
+                            _numUserWaits = i - 1;
+                            break;
+                        }
+                    }
+                }
+
+                // Recalculate Timeout
+                int timeout = Timeout.Infinite;
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (timeout == Timeout.Infinite)
```

Comment:
The user can register a wait with an infinite timeout, so we need to keep this check in even if we start calculations at `i=1`.